drivers/eeprom/eeprom_mb85rsxx.c

```diff
@@ -0,0 +1,318 @@
+#define EEPROM_MB85RSXX_STATUS_WEL  BIT(1) /* Write Enable Latch			  (RO) */
+
+/* Fujitsu manufacturer ID (2 bytes) */
+#define EEPROM_MB85RSXX_MAN_ID		0x04U
+#define EEPROM_MB85RSXX_CON_CODE	0x7FU
+
+/*
+ * MB85RSXX product ID (2 bytes); first byte provides memory size, so let's use a mask later when
+ * checking it
+ */
+#define EEPROM_MB85RSXX_PROD_ID		0x20U
+#define EEPROM_MB85RSXX_PROD_ID2	0x03U
+#define EEPROM_MB85RSXX_PROD_MASK	GENMASK(7, 5)
+
+struct eeprom_mb85rsxx_config {
+	struct spi_dt_spec spi;
+	size_t size;
+	bool readonly;
+};
+
+struct eeprom_mb85rsxx_data {
+	struct k_mutex lock;
+};
+
+static int eeprom_mb85rsxx_read(const struct device *dev, off_t offset, void *buf, size_t len)
+{
+	const struct eeprom_mb85rsxx_config *config = dev->config;
+	struct eeprom_mb85rsxx_data *data = dev->data;
+	uint8_t cmd[4] = {EEPROM_MB85RSXX_READ, 0, 0, 0};
+	uint8_t *paddr = &cmd[1];
+	int err;
+
+	if (offset + len > config->size) {
+		LOG_ERR("attempt to read past device boundary");
+		return -EINVAL;
+	}
+
+	if (!len) {
+		return 0;
+	}
+
+	/* Populate address in command */
+	*paddr++ = (offset >> 16);
+	*paddr++ = (offset >> 8);
+	*paddr++ = offset;
+
+	const struct spi_buf tx_buf = {
+		.buf = cmd,
+		.len = sizeof(cmd),
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1,
+	};
+	const struct spi_buf rx_bufs[2] = {
+		{
+			.buf = NULL,
+			.len = sizeof(cmd),
+		},
+		{
+			.buf = buf,
+			.len = len,
+		},
+	};
+	const struct spi_buf_set rx = {
+		.buffers = rx_bufs,
+		.count = ARRAY_SIZE(rx_bufs),
+	};
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	err = spi_transceive_dt(&config->spi, &tx, &rx);
+
+	k_mutex_unlock(&data->lock);
+
+	if (err < 0) {
+		LOG_ERR("failed to read FRAM (err %d)", err);
+	}
+
+	return err;
+}
+
+static int eeprom_mb85rsxx_wren(const struct device *dev)
+{
+	const struct eeprom_mb85rsxx_config *config = dev->config;
+	uint8_t cmd = EEPROM_MB85RSXX_WREN;
+	const struct spi_buf tx_buf = {
+		.buf = &cmd,
+		.len = sizeof(cmd),
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1,
+	};
+
+	return spi_write_dt(&config->spi, &tx);
+}
+
+static int eeprom_mb85rsxx_wrdi(const struct device *dev)
+{
+	const struct eeprom_mb85rsxx_config *config = dev->config;
+	uint8_t cmd = EEPROM_MB85RSXX_WRDI;
+	const struct spi_buf tx_buf = {
+		.buf = &cmd,
+		.len = sizeof(cmd),
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1,
+	};
+
+	return spi_write_dt(&config->spi, &tx);
+}
+
+static int eeprom_mb85rsxx_write(const struct device *dev, off_t offset, const void *buf,
+				  size_t len)
+{
+	const struct eeprom_mb85rsxx_config *config = dev->config;
+	struct eeprom_mb85rsxx_data *data = dev->data;
+	uint8_t cmd[4] = {EEPROM_MB85RSXX_WRITE, 0, 0, 0};
+	uint8_t *paddr = &cmd[1];
+	int err;
+
+	if (config->readonly) {
+		LOG_ERR("attempt to write to read-only device");
+		return -EACCES;
+	}
```

Comment:
Nit. It can go in the next PR: I have seen that you have defined RDSR and BP0 and BP1 bits, you are not giving a way to modify them, nevertheless the memory may come with these bits already set, in which case write to regions covered by these bits will fail  - as they are readonly.

Whether the RDSR is done in init or before every write is up to discussion, but write should immediately return error when bits are set; the FRAM will not write anything in are with bit set, but there will be no error in return.